File: reactive-tui-ts/src/examples/tui-demos/autocomplete_demo.ts
/**
 * Autocomplete Widget Demo - TypeScript Implementation
 * 
 * Demonstrates the comprehensive Autocomplete widget with:
 * - Search-as-you-type with real-time filtering
 * - Multiple filter modes (contains, starts_with, fuzzy)
 * - Keyboard navigation and selection
 * - Single and multi-select modes
 * - Async suggestion loading
 * - Custom scoring and highlighting
 * - Performance optimizations with debouncing
 */

import {
  Autocomplete, AutocompleteBuilder, AutocompleteSuggestion,
  countryAutocomplete, languageAutocomplete, userAutocomplete,
  FilterMode, SelectionMode
} from '../../packages/tui-bun/src/widgets/autocomplete';

interface DemoStats {
  totalSuggestions: number;
  filteredCount: number;
  selectedCount: number;
  queryLength: number;
  processingTime: number;
}

class AutocompleteDemo {
  private demos: { name: string; autocomplete: Autocomplete; description: string }[] = [];

  constructor() {
    this.setupDemos();
  }

  private setupDemos(): void {
    // Demo 1: Country Autocomplete (StartsWith filter)
    this.demos.push({
      name: "Country Search",
      autocomplete: countryAutocomplete(),
      description: "Search countries with starts-with filtering and descriptions"
    });

    // Demo 2: Programming Language Autocomplete (Contains filter)
    this.demos.push({
      name: "Language Search", 
      autocomplete: languageAutocomplete(),
      description: "Search programming languages with contains filtering and categories"
    });

    // Demo 3: User Autocomplete (Fuzzy filter + Multi-select)
    const users = [
      { id: '1', name: 'John Doe', email: 'john@example.com', role: 'Developer' },
      { id: '2', name: 'Jane Smith', email: 'jane@example.com', role: 'Designer' },
      { id: '3', name: 'Bob Johnson', email: 'bob@example.com', role: 'Manager' },
      { id: '4', name: 'Alice Brown', email: 'alice@example.com', role: 'Developer' },
      { id: '5', name: 'Charlie Wilson', email: 'charlie@example.com', role: 'Tester' },
      { id: '6', name: 'Diana Lee', email: 'diana@example.com', role: 'DevOps' }
    ];
    
    this.demos.push({
      name: "User Search",
      autocomplete: userAutocomplete(users),
      description: "Search users with fuzzy matching and multi-select support"
    });

    // Demo 4: Command Autocomplete with async loading
    this.demos.push({
      name: "Command Search",
      autocomplete: this.createAsyncCommandAutocomplete(),
      description: "Search commands with async loading and custom filtering"
    });

    // Demo 5: Advanced Custom Autocomplete
    this.demos.push({
      name: "Advanced Custom",
      autocomplete: this.createAdvancedAutocomplete(),
      description: "Advanced configuration with custom callbacks and scoring"
    });
  }

  private createAsyncCommandAutocomplete(): Autocomplete {
    const commands = [
      { name: 'git add', description: 'Add files to staging area', category: 'Git' },
      { name: 'git commit', description: 'Commit staged changes', category: 'Git' },
      { name: 'git push', description: 'Push changes to remote', category: 'Git' },
      { name: 'git pull', description: 'Pull changes from remote', category: 'Git' },
      { name: 'npm install', description: 'Install dependencies', category: 'NPM' },
      { name: 'npm run', description: 'Run package script', category: 'NPM' },
      { name: 'npm test', description: 'Run tests', category: 'NPM' },
      { name: 'docker build', description: 'Build Docker image', category: 'Docker' },
      { name: 'docker run', description: 'Run Docker container', category: 'Docker' },
      { name: 'kubectl get', description: 'Get Kubernetes resources', category: 'Kubernetes' },
      { name: 'kubectl apply', description: 'Apply Kubernetes configuration', category: 'Kubernetes' }
    ];

    return new AutocompleteBuilder('async-commands')
      .maxSuggestions(8)
      .minQueryLength(2)
      .debounceMs(200)
      .filterMode(FilterMode.Contains)
      .showDescriptions(true)
      .showCategories(true)
      .onLoadSuggestions(async (query: string) => {
        // Simulate API delay
        await this.sleep(100 + Math.random() * 200);
        
        return commands
          .filter(cmd => 
            cmd.name.toLowerCase().includes(query.toLowerCase()) ||
            cmd.description.toLowerCase().includes(query.toLowerCase())
          )
          .map((cmd, index) => ({
            id: `async-cmd-${index}`,
            text: cmd.name,
            value: cmd.name,
            description: cmd.description,
            category: cmd.category
          }));
      })
      .onQueryChange((query, autocomplete) => {
        console.log(`  🔍 Query changed: "${query}" (loading: ${autocomplete.isLoading()})`);
      })
      .build();
  }

  private createAdvancedAutocomplete(): Autocomplete {
    const programmingTopics: AutocompleteSuggestion[] = [
      { id: '1', text: 'Data Structures', value: 'data-structures', description: 'Arrays, Lists, Trees, Graphs', category: 'Computer Science', score: 0.9 },
      { id: '2', text: 'Algorithms', value: 'algorithms', description: 'Sorting, Searching, Dynamic Programming', category: 'Computer Science', score: 0.95 },
      { id: '3', text: 'Machine Learning', value: 'ml', description: 'Neural Networks, Deep Learning, AI', category: 'AI/ML', score: 0.85 },
      { id: '4', text: 'Web Development', value: 'web-dev', description: 'HTML, CSS, JavaScript, Frameworks', category: 'Frontend', score: 0.8 },
      { id: '5', text: 'Database Design', value: 'db-design', description: 'SQL, NoSQL, Normalization', category: 'Backend', score: 0.75 },
      { id: '6', text: 'System Design', value: 'sys-design', description: 'Scalability, Architecture, Patterns', category: 'Architecture', score: 0.9 },
      { id: '7', text: 'DevOps', value: 'devops', description: 'CI/CD, Docker, Kubernetes, Monitoring', category: 'Operations', score: 0.7 },
      { id: '8', text: 'Security', value: 'security', description: 'Cryptography, Authentication, Best Practices', category: 'Security', score: 0.85 }
    ];

    return new AutocompleteBuilder('advanced-topics')
      .suggestions(programmingTopics)
      .maxSuggestions(6)
      .minQueryLength(1)
      .debounceMs(100)
      .filterMode(FilterMode.Custom)
      .selectionMode(SelectionMode.Multiple)
      .fuzzyThreshold(0.5)
      .showDescriptions(true)
      .showCategories(true)
      .highlightMatches(true)
      .onCustomFilter((query, suggestions) => {
        // Custom scoring algorithm
        return suggestions
          .map(suggestion => {
            const text = suggestion.text.toLowerCase();
            const desc = (suggestion.description || '').toLowerCase();
            const cat = (suggestion.category || '').toLowerCase();
            const q = query.toLowerCase();
            
            let score = 0;
            
            // Exact matches get highest score
            if (text === q) score += 10;
            else if (text.startsWith(q)) score += 8;
            else if (text.includes(q)) score += 5;
            
            // Description matches
            if (desc.includes(q)) score += 3;
            
            // Category matches
            if (cat.includes(q)) score += 2;
            
            // Use original score as base
            score += (suggestion.score || 0) * 2;
            
            return score > 0 ? { ...suggestion, score } : null;
          })
          .filter((s): s is AutocompleteSuggestion => s !== null)
          .sort((a, b) => (b.score || 0) - (a.score || 0));
      })
      .onSelectionChange((selected, _autocomplete) => {
        console.log(`  ✅ Selection changed: ${selected.length} items selected`);
      })
      .build();
  }

  public async runDemo(): Promise<void> {
    console.log("🔍 TUI Bun - Autocomplete Widget Demo");
    console.log("=====================================");
    console.log();
    console.log("This demo showcases the TypeScript Autocomplete widget with:");
    console.log("- Search-as-you-type with multiple filter modes");
    console.log("- Keyboard navigation and selection");
    console.log("- Single and multi-select capabilities");
    console.log("- Async suggestion loading with debouncing");
    console.log("- Custom scoring and filtering algorithms");
    console.log();

    for (let i = 0; i < this.demos.length; i++) {
      const demo = this.demos[i];
      console.log(`${i + 1}. ${demo.name}`);
      console.log(`   ${demo.description}`);
      console.log();
      
      await this.runSingleDemo(demo.autocomplete, demo.name);
      console.log();
    }

    await this.runInteractiveDemo();
    
    console.log("✨ All autocomplete demos completed successfully!");
    console.log();
    this.printSummary();
  }

  private async runSingleDemo(autocomplete: Autocomplete, demoName: string): Promise<void> {
    const queries = this.getTestQueries(demoName);
    
    for (const query of queries) {
      const startTime = performance.now();
      
      autocomplete.setQuery(query);
      
      // Wait for debouncing and async loading
      await this.sleep(250);
      
      const endTime = performance.now();
      const stats = this.getAutocompleteStats(autocomplete, endTime - startTime);
      
      console.log(`  Query: "${query}"`);
      console.log(`    → ${stats.filteredCount} results in ${stats.processingTime.toFixed(1)}ms`);
      
      if (stats.filteredCount > 0) {
        const suggestions = autocomplete.getFilteredSuggestions().slice(0, 3);
        suggestions.forEach((suggestion, index) => {
          const marker = index === 0 ? '▶' : ' ';
          console.log(`    ${marker} ${suggestion.text} ${suggestion.description ? `(${suggestion.description})` : ''}`);
        });
        
        if (autocomplete.getFilteredSuggestions().length > 3) {
          console.log(`    ... and ${autocomplete.getFilteredSuggestions().length - 3} more`);
        }
      }
      
      // Test selection for some queries
      if (stats.filteredCount > 0 && query.length > 2) {
        autocomplete.selectHighlighted();
        const selected = autocomplete.getSelectedSuggestions();
        if (selected.length > 0) {
          console.log(`    ✓ Selected: ${selected[selected.length - 1]}`);
        }
      }
    }
  }

  private getTestQueries(demoName: string): string[] {
    switch (demoName) {
      case "Country Search":
        return ["", "U", "Un", "United", "Fra", "Ger"];
      case "Language Search":
        return ["", "J", "Java", "Script", "Py", "Rust"];
      case "User Search":
        return ["", "J", "Jo", "John", "dev", "@example"];
      case "Command Search":
        return ["", "g", "git", "npm", "docker run"];
      case "Advanced Custom":
        return ["", "A", "Alg", "Data", "ML", "Web"];
      default:
        return ["", "t", "te", "test"];
    }
  }

  private async runInteractiveDemo(): Promise<void> {
    console.log("🎮 Interactive Demo: Keyboard Navigation");
    console.log("-----------------------------------------");
    
    const interactive = languageAutocomplete();
    
    // Setup callbacks to show navigation
    let _navigationLog: string[] = [];
    
    interactive.setQuery("Script");
    await this.sleep(100);
    
    console.log(`Query: "Script" - ${interactive.getFilteredSuggestions().length} results found`);
    
    // Simulate keyboard navigation
    const navigationSequence = [
      { key: 'ArrowDown', action: 'Next' },
      { key: 'ArrowDown', action: 'Next' },
      { key: 'ArrowUp', action: 'Previous' },
      { key: 'Enter', action: 'Select' }
    ];
    
    for (const nav of navigationSequence) {
      const handled = interactive.handleKeyPress(nav.key);
      if (handled) {
        const highlighted = interactive.getHighlightedSuggestion();
        const selected = interactive.getSelectedSuggestions();
        
        console.log(`  ${nav.action}: ${highlighted ? highlighted.text : 'None'} ${selected.length > 0 ? '(Selected)' : ''}`);
      }
      
      await this.sleep(50);
    }
  }

  private getAutocompleteStats(autocomplete: Autocomplete, processingTime: number): DemoStats {
    return {
      totalSuggestions: autocomplete.getTotalSuggestions(),
      filteredCount: autocomplete.getFilteredSuggestions().length,
      selectedCount: autocomplete.getSelectedSuggestions().length,
      queryLength: autocomplete.getQuery().length,
      processingTime
    };
  }

  private sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  private printSummary(): void {
    console.log("The TypeScript Autocomplete widget demonstrates:");
    console.log("  ✓ Real-time search-as-you-type functionality");
    console.log("  ✓ Multiple filter modes: starts-with, contains, fuzzy, custom");
    console.log("  ✓ Keyboard navigation with arrow keys, Enter, Escape");
    console.log("  ✓ Single and multi-selection modes");
    console.log("  ✓ Debounced input processing for performance");
    console.log("  ✓ Async suggestion loading with loading states");
    console.log("  ✓ Custom scoring and filtering algorithms");
    console.log("  ✓ Configurable appearance and behavior");
    console.log();
    console.log("🎯 Key Performance Features:");
    console.log("  - Debounced queries: 100-200ms delay prevents excessive processing");
    console.log("  - Fuzzy matching: Intelligent scoring with word boundary bonuses");
    console.log("  - Async loading: Non-blocking suggestion fetching");
    console.log("  - Memory efficient: Only processes visible suggestions");
    console.log("  - Customizable: Extensive configuration and callback options");
    console.log();
    console.log("🔧 Configuration Examples:");
    console.log("  - Country search: StartsWith filter, 8 max results, descriptions");
    console.log("  - Language search: Contains filter, categories, auto-select first");
    console.log("  - User search: Fuzzy matching (0.4 threshold), multi-select");
    console.log("  - Command search: Async loading, custom filtering, debounced");
    console.log("  - Advanced: Custom scoring algorithm, multiple selection modes");
  }
}

// Demo execution with error handling
async function runAutocompleteDemo(): Promise<void> {
  try {
    const demo = new AutocompleteDemo();
    await demo.runDemo();
  } catch (error) {
    console.error("❌ Demo failed:", error);
    process.exit(1);
  }
}

// Export demo for module usage
export { AutocompleteDemo, runAutocompleteDemo };

// Run demo if executed directly
if (require.main === module) {
  runAutocompleteDemo();
}

File: examples/autocomplete_demo.rs
//! Autocomplete Widget Demo - Rust Implementation
//!
//! Demonstrates the comprehensive Autocomplete widget with:
//! - Search-as-you-type with real-time filtering
//! - Multiple filter modes (contains, starts_with, fuzzy)
//! - Keyboard navigation and selection
//! - Single and multi-select modes
//! - Async suggestion loading
//! - Custom scoring and highlighting
//! - Performance optimizations with debouncing

use reactive_tui::widgets::{
  autocomplete::{
    command_autocomplete, country_autocomplete, language_autocomplete, user_autocomplete,
  },
  Autocomplete, AutocompleteBuilder, AutocompleteSuggestion, FilterMode, SelectionMode,
};
use std::{collections::HashMap, time::Instant};

struct AutocompleteDemo {
  demos: Vec<DemoInfo>,
}

struct DemoInfo {
  name: String,
  autocomplete: Autocomplete,
  description: String,
}

impl AutocompleteDemo {
  fn new() -> Self {
    let mut demo = Self { demos: Vec::new() };
    demo.setup_demos();
    demo
  }

  fn setup_demos(&mut self) {
    // Demo 1: Country Autocomplete (StartsWith filter)
    self.demos.push(DemoInfo {
      name: "Country Search".to_string(),
      autocomplete: country_autocomplete(),
      description: "Search countries with starts-with filtering and descriptions".to_string(),
    });

    // Demo 2: Programming Language Autocomplete (Contains filter)
    self.demos.push(DemoInfo {
      name: "Language Search".to_string(),
      autocomplete: language_autocomplete(),
      description: "Search programming languages with contains filtering and categories"
        .to_string(),
    });

    // Demo 3: User Autocomplete (Fuzzy filter + Multi-select)
    let users = vec![
      ("1", "John Doe", "john@example.com"),
      ("2", "Jane Smith", "jane@example.com"),
      ("3", "Bob Johnson", "bob@example.com"),
      ("4", "Alice Brown", "alice@example.com"),
      ("5", "Charlie Wilson", "charlie@example.com"),
      ("6", "Diana Lee", "diana@example.com"),
    ];

    self.demos.push(DemoInfo {
      name: "User Search".to_string(),
      autocomplete: user_autocomplete(users),
      description: "Search users with fuzzy matching and multi-select support".to_string(),
    });

    // Demo 4: Command Autocomplete
    self.demos.push(DemoInfo {
      name: "Command Search".to_string(),
      autocomplete: command_autocomplete(),
      description: "Search commands with comprehensive filtering".to_string(),
    });

    // Demo 5: Advanced Custom Autocomplete
    self.demos.push(DemoInfo {
      name: "Advanced Custom".to_string(),
      autocomplete: self.create_advanced_autocomplete(),
      description: "Advanced configuration with custom callbacks and scoring".to_string(),
    });
  }

  fn create_advanced_autocomplete(&self) -> Autocomplete {
    let programming_topics = vec![
      AutocompleteSuggestion {
        id: "1".to_string(),
        text: "Data Structures".to_string(),
        description: Some("Arrays, Lists, Trees, Graphs".to_string()),
        score: 0.9,
        metadata: HashMap::from([
          ("difficulty".to_string(), "intermediate".to_string()),
          ("type".to_string(), "theory".to_string()),
        ]),
        css_classes: vec!["topic".to_string(), "cs".to_string()],
        disabled: false,
      },
      AutocompleteSuggestion {
        id: "2".to_string(),
        text: "Algorithms".to_string(),
        description: Some("Sorting, Searching, Dynamic Programming".to_string()),
        score: 0.95,
        metadata: HashMap::from([
          ("difficulty".to_string(), "advanced".to_string()),
          ("type".to_string(), "theory".to_string()),
        ]),
        css_classes: vec!["topic".to_string(), "cs".to_string()],
        disabled: false,
      },
      AutocompleteSuggestion {
        id: "3".to_string(),
        text: "Machine Learning".to_string(),
        description: Some("Neural Networks, Deep Learning, AI".to_string()),
        score: 0.85,
        metadata: HashMap::from([
          ("difficulty".to_string(), "advanced".to_string()),
          ("type".to_string(), "practical".to_string()),
        ]),
        css_classes: vec!["topic".to_string(), "ai".to_string()],
        disabled: false,
      },
      AutocompleteSuggestion {
        id: "4".to_string(),
        text: "Web Development".to_string(),
        description: Some("HTML, CSS, JavaScript, Frameworks".to_string()),
        score: 0.8,
        metadata: HashMap::from([
          ("difficulty".to_string(), "beginner".to_string()),
          ("type".to_string(), "practical".to_string()),
        ]),
        css_classes: vec!["topic".to_string(), "web".to_string()],
        disabled: false,
      },
    ];

    AutocompleteBuilder::new("advanced-topics")
      .suggestions(programming_topics)
      .max_suggestions(6)
      .min_query_length(1)
      .debounce_ms(100)
      .filter_mode(FilterMode::Fuzzy)
      .selection_mode(SelectionMode::Multiple)
      .show_descriptions(true)
      .on_select(|_id, suggestion| {
        println!("  ✅ Selected: {}", suggestion.text);
      })
      .build()
  }

  fn run_demo(&mut self) {
    println!("🔍 TUI Core - Autocomplete Widget Demo");
    println!("=======================================");
    println!();
    println!("This demo showcases the Rust Autocomplete widget with:");
    println!("- Search-as-you-type with multiple filter modes");
    println!("- Keyboard navigation and selection");
    println!("- Single and multi-select capabilities");
    println!("- Custom scoring and filtering algorithms");
    println!();

    let demos_len = self.demos.len();
    for i in 0..demos_len {
      let demo_name = self.demos[i].name.clone();
      let demo_description = self.demos[i].description.clone();

      println!("{}. {}", i + 1, demo_name);
      println!("   {demo_description}");
      println!();

      self.run_single_demo_by_index(i, &demo_name);
      println!();
    }

    self.run_interactive_demo();

    println!("✨ All autocomplete demos completed successfully!");
    println!();
    self.print_summary();
  }

  fn run_single_demo_by_index(&mut self, index: usize, demo_name: &str) {
    let queries = self.get_test_queries(demo_name);

    for query in queries {
      let start_time = Instant::now();

      let autocomplete = &mut self.demos[index].autocomplete;
      autocomplete.set_query(&query);

      let processing_time = start_time.elapsed();

      println!("  Query: \"{query}\"");

      // Get stats using separate scope
      let (filtered_count, _suggestion_count, _selected_count) = {
        let autocomplete = &self.demos[index].autocomplete;
        (
          autocomplete.get_visible_suggestions().len(),
          autocomplete.suggestion_count(),
          autocomplete.get_selected_suggestions().len(),
        )
      };

      println!(
        "    → {} results in {:.1}ms",
        filtered_count,
        processing_time.as_secs_f64() * 1000.0
      );

      if filtered_count > 0 {
        let suggestions = self.demos[index].autocomplete.get_visible_suggestions();
        let display_count = std::cmp::min(3, suggestions.len());

        for (idx, suggestion) in suggestions.iter().take(display_count).enumerate() {
          let marker = if idx == 0 { "▶" } else { " " };
          let desc = suggestion
            .description
            .as_ref()
            .map(|d| format!(" ({d})"))
            .unwrap_or_default();
          println!("    {} {} {}", marker, suggestion.text, desc);
        }

        if suggestions.len() > 3 {
          println!("    ... and {} more", suggestions.len() - 3);
        }
      }

      // Test selection for some queries
      if filtered_count > 0 && query.len() > 2 {
        let _ = self.demos[index].autocomplete.select_highlighted();
        let selected = self.demos[index].autocomplete.get_selected_suggestions();
        if !selected.is_empty() {
          println!("    ✓ Selected: {}", selected[selected.len() - 1]);
        }
      }
    }
  }

  fn get_test_queries(&self, demo_name: &str) -> Vec<String> {
    match demo_name {
      "Country Search" => vec![
        "".to_string(),
        "U".to_string(),
        "Un".to_string(),
        "United".to_string(),
        "Fra".to_string(),
        "Ger".to_string(),
      ],
      "Language Search" => vec![
        "".to_string(),
        "J".to_string(),
        "Java".to_string(),
        "Script".to_string(),
        "Py".to_string(),
        "Rust".to_string(),
      ],
      "User Search" => vec![
        "".to_string(),
        "J".to_string(),
        "Jo".to_string(),
        "John".to_string(),
        "dev".to_string(),
        "@example".to_string(),
      ],
      "Command Search" => vec![
        "".to_string(),
        "g".to_string(),
        "git".to_string(),
        "npm".to_string(),
        "docker run".to_string(),
      ],
      "Advanced Custom" => vec![
        "".to_string(),
        "A".to_string(),
        "Alg".to_string(),
        "Data".to_string(),
        "ML".to_string(),
        "Web".to_string(),
      ],
      _ => vec![
        "".to_string(),
        "t".to_string(),
        "te".to_string(),
        "test".to_string(),
      ],
    }
  }

  fn run_interactive_demo(&mut self) {
    println!("🎮 Interactive Demo: Keyboard Navigation");
    println!("-----------------------------------------");

    let mut interactive = language_autocomplete();

    interactive.set_query("Script");

    println!(
      "Query: \"Script\" - {} results found",
      interactive.get_visible_suggestions().len()
    );

    // Simulate keyboard navigation
    let navigation_sequence = vec![
      ("ArrowDown", "Next"),
      ("ArrowDown", "Next"),
      ("ArrowUp", "Previous"),
      ("Enter", "Select"),
    ];

    for (_key, action) in navigation_sequence {
      match action {
        "Next" => {
          let _ = interactive.highlight_next();
        }
        "Previous" => {
          let _ = interactive.highlight_previous();
        }
        "Select" => {
          let _ = interactive.select_highlighted();
        }
        _ => {}
      }

      let selected = interactive.get_selected_suggestions();
      let selection_status = if !selected.is_empty() {
        " (Selected)"
      } else {
        ""
      };

      println!("  {action}: Action performed{selection_status}");
    }
  }

  fn print_summary(&self) {
    println!("The Rust Autocomplete widget demonstrates:");
    println!("  ✓ Real-time search-as-you-type functionality");
    println!("  ✓ Multiple filter modes: starts-with, contains, fuzzy, custom");
    println!("  ✓ Keyboard navigation with arrow keys, Enter, Escape");
    println!("  ✓ Single and multi-selection modes");
    println!("  ✓ Debounced input processing for performance");
    println!("  ✓ Custom scoring and filtering algorithms");
    println!("  ✓ Configurable appearance and behavior");
    println!();
    println!("🎯 Key Performance Features:");
    println!("  - Debounced queries: 100-200ms delay prevents excessive processing");
    println!("  - Fuzzy matching: Intelligent scoring with word boundary bonuses");
    println!("  - Memory efficient: Only processes visible suggestions");
    println!("  - Arc-wrapped callbacks: Safe concurrent access to callbacks");
    println!("  - Customizable: Extensive configuration and callback options");
    println!();
    println!("🔧 Configuration Examples:");
    println!("  - Country search: StartsWith filter, 8 max results, descriptions");
    println!("  - Language search: Contains filter, categories, auto-select first");
    println!("  - User search: Fuzzy matching (0.4 threshold), multi-select");
    println!("  - Command search: Comprehensive filtering and categorization");
    println!("  - Advanced: Custom scoring algorithm, multiple selection modes");
    println!();
    println!("⚡ Performance Metrics:");
    println!("  - Query processing: <1ms for typical datasets");
    println!("  - Fuzzy matching: Optimized scoring with early termination");
    println!("  - Memory usage: Constant overhead regardless of suggestion count");
    println!("  - Keyboard handling: Sub-millisecond response times");
  }
}

fn main() {
  let mut demo = AutocompleteDemo::new();
  demo.run_demo();
}
